File: apps/promotion-service/src/main/java/com/booster/promotionservice/domain/entity/Coupon.java
package com.booster.promotionservice.domain.entity;

import com.booster.common.SnowflakeGenerator;
import com.booster.storage.db.core.BaseEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Table(name = "coupon")
public class Coupon extends BaseEntity {

    @Id
    private Long id; // Auto Increment 아님!

    @Column(nullable = false)
    private String title;

    @Column(nullable = false)
    private Integer totalQuantity;

    @Column(nullable = false)
    private Integer issuedQuantity = 0; // 통계 목적 (Redis와 별개)

    @Column(nullable = false)
    private Integer discountAmount;

    @Column(nullable = false)
    private LocalDateTime startAt;

    @Column(nullable = false)
    private LocalDateTime endAt;

    // 생성자에서 ID 주입
    public Coupon(String title, Integer totalQuantity, Integer discountAmount, LocalDateTime startAt, LocalDateTime endAt) {
        this.id = SnowflakeGenerator.nextId(); // Snowflake ID 생성
        this.title = title;
        this.totalQuantity = totalQuantity;
        this.discountAmount = discountAmount;
        this.startAt = startAt;
        this.endAt = endAt;
    }

    // 재고 증가 메서드 (동시성 처리 주의 필요하지만, 여기선 단순 통계용 업데이트)
    public void increaseIssuedQuantity() {
        this.issuedQuantity++;
    }

}

File: apps/kotlin/shopping-service/src/main/kotlin/com/booster/kotlin/shoppingservice/coupon/domain/Coupon.kt
package com.booster.kotlin.shoppingservice.coupon.domain

import com.booster.kotlin.shoppingservice.common.exception.ErrorCode
import com.booster.kotlin.shoppingservice.coupon.exception.CouponException
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import jakarta.persistence.EntityListeners
import java.time.LocalDateTime

@Entity
@Table(name = "coupons")
@EntityListeners(AuditingEntityListener::class)
class Coupon(
    @Column(nullable = false, unique = true) val code: String,
    @Column(nullable = false) val name: String,
    @Enumerated(EnumType.STRING) @Column(nullable = false) val couponType: CouponType,
    @Column(nullable = false) val discountValue: Long,      // FIXED: 원, RATE: % (1~100)
    @Column val maxDiscountAmount: Long?,                   // RATE 전용 할인 상한
    @Column(nullable = false) val minOrderAmount: Long,     // 최소 주문 금액
    @Column(nullable = false) val startAt: LocalDateTime,
    @Column(nullable = false) val endAt: LocalDateTime,
) {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: CouponStatus = CouponStatus.ACTIVE
        private set

    @CreatedDate
    @Column(updatable = false)
    lateinit var createdAt: LocalDateTime

    @OneToMany(mappedBy = "coupon", cascade = [CascadeType.ALL], orphanRemoval = true)
    val productTargets: MutableList<CouponProductTarget> = mutableListOf()

    @OneToMany(mappedBy = "coupon", cascade = [CascadeType.ALL], orphanRemoval = true)
    val categoryTargets: MutableList<CouponCategoryTarget> = mutableListOf()

    fun isAvailable(): Boolean =
        status == CouponStatus.ACTIVE && LocalDateTime.now() in startAt..endAt

    fun deactivate() {
        status = CouponStatus.INACTIVE
    }

    /**
     * 주문 금액에 대한 실제 할인 금액 계산.
     * - FIXED: discountValue 원 (주문 금액 초과 불가)
     * - RATE: 주문 금액 × discountValue% (maxDiscountAmount 상한 적용)
     */
    fun calculateDiscount(orderAmount: Long): Long {
        if (orderAmount < minOrderAmount) {
            throw CouponException(ErrorCode.COUPON_MIN_ORDER_AMOUNT_NOT_MET)
        }
        return when (couponType) {
            CouponType.FIXED -> discountValue.coerceAtMost(orderAmount)
            CouponType.RATE -> {
                val discount = orderAmount * discountValue / 100
                maxDiscountAmount?.let { discount.coerceAtMost(it) } ?: discount
            }
        }
    }

    companion object {
        fun create(
            code: String,
            name: String,
            couponType: CouponType,
            discountValue: Long,
            maxDiscountAmount: Long?,
            minOrderAmount: Long,
            startAt: LocalDateTime,
            endAt: LocalDateTime,
        ) = Coupon(code, name, couponType, discountValue, maxDiscountAmount, minOrderAmount, startAt, endAt)
    }
}